components/AttackerPresets.js

```diff
@@ -0,0 +1,257 @@
+				player.selectPrayer(prayer);
+			}
+		}
+
+		if (Array.isArray(preset.boosts)) {
+			player.clearBoosts();
+			for (const boost of preset.boosts) {
+				player.addBoost(boost);
+			}
+		}
+
+		props.setPlayer(player.minimize());
+		props.setTab(0);
+	};
+
+	const removeSetup = (evt, name) => {
+		evt.preventDefault();
+
+		const index = customPresets.findIndex(i => i.name === name);
+		if (index === -1) {
+			return;
+		}
+
+		const result = globalThis.confirm(`Really remove preset "${name}"?`);
+		if (!result) {
+			return;
+		}
+
+		const copy = [...customPresets];
+		copy.splice(index, 1);
+
+		setCustomPresets(copy);
+		mappedCustomPresets.delete(name);
+		renderedCustomPresets.delete(name);
+	};
+
+	const saveCustomPreset = (evt, player) => {
+		if (evt.key !== "Enter") {
+			return;
+		}
+
+		const input = saveCustomPresetInput.ref.current;
+		const name = input.value.trim();
+		if (!name) {
+			globalThis.alert("Invalid preset name!");
+			return;
+		}
+
+		const itemIds = Object.values(player.equipment).map(i => i.id).filter(Boolean);
+		let preset = customPresets.find(i => i.name === name);
+		if (preset) {
+			preset.equipmentIdentifiers = itemIds;
+			preset.created = Date.now();
+		}
+		else {
+			preset = {
+				name,
+				equipmentIdentifiers: itemIds,
+				created: Date.now()
+			};
+
+			customPresets.push(preset);
+		}
+
+		setCustomPresets([...customPresets]);
+		input.value = "";
+
+		mappedCustomPresets.set(preset.name, mapPreset(preset));
+		forceUpdate();
+	};
+
+	const clearCustomPresets = () => {
+		const result = globalThis.confirm("Really remove all of your custom presets?");
+		if (result) {
+			setCustomPresets([]);
+		}
+	};
+
+	if (customPresets.length === 0) {
+		mappedCustomPresets.clear();
+		renderedCustomPresets.clear();
+	}
+	else {
+		for (const customPreset of customPresets) {
+			const { name } = customPreset;
+			const isMapped = mappedCustomPresets.has(name);
+			if (!isMapped) {
+				mappedCustomPresets.set(name, mapPreset(customPreset));
+			}
+
+			const existing = renderedCustomPresets.get(name);
+			if (existing) {
+				continue;
+			}
+
+			renderedCustomPresets.set(name, (
+				<button
+					onClick={() => changeSetup(mappedCustomPresets.get(name))}
+					onContextMenu={(evt) => removeSetup(evt, name)}
+				>
+					{name}
+				</button>
+			));
+		}
+
+		for (const [name] of renderedCustomPresets) {
+			const existing = customPresets.some(i => i.name === name);
+			if (existing) {
+				continue;
+			}
+
+			renderedCustomPresets.delete(name);
+		}
+	}
+
+	const customPresetSeparator = (renderedCustomPresets.size > 0)
+		? <><br/><h4>Custom presets (right click to remove)</h4></>
+		: null;
+
+	const removeAllCustomPresetsElement = (renderedCustomPresets.size > 0)
+		? <><br/><button onClick={clearCustomPresets}>Remove all custom presets</button></>
+		: null;
+
+	return (
+		<div className="highlight-section flex-container-vertical">
+			<h3>Apply setup preset</h3>
+			{...elements}
```

Comment:
i dislike the way you split the UI into variables. (personal preference ofc)
but i would either turn these into their own actual mini components OR just inline them to give the html a visible structure